src/effects/effectchainslot.cpp

```diff
@@ -113,190 +136,199 @@ EffectChainSlot::~EffectChainSlot() {
-        } else {
-            m_pEffectChain->disableForInputChannel(pChannelInfo->handle_group);
-        }
-    }
+const QList<EffectPointer>& EffectChainSlot::effects() const {
+    return m_effects;
+}
+
+EngineEffectChain* EffectChainSlot::getEngineEffectChain() {
+    return m_pEngineEffectChain;
+}
+
+void EffectChainSlot::sendParameterUpdate() {
+    EffectsRequest* pRequest = new EffectsRequest();
+    pRequest->type = EffectsRequest::SET_EFFECT_CHAIN_PARAMETERS;
+    pRequest->pTargetChain = m_pEngineEffectChain;
+    pRequest->SetEffectChainParameters.enabled = m_pControlChainEnabled->get();
+    pRequest->SetEffectChainParameters.mix_mode = static_cast<EffectChainMixMode>(
+                                                    static_cast<int>(m_pControlChainMixMode->get()));
+    pRequest->SetEffectChainParameters.mix = m_pControlChainMix->get();
+    m_pEffectsManager->writeRequest(pRequest);
+}
+
+QString EffectChainSlot::id() const {
+    return m_id;
 }
 
-EffectChainPointer EffectChainSlot::getEffectChain() const {
-    return m_pEffectChain;
+double EffectChainSlot::getSuperParameter() const {
+    return m_pControlChainSuperParameter->get();
+}
+
+void EffectChainSlot::setSuperParameter(double value, bool force) {
+    m_pControlChainSuperParameter->set(value);
+    slotControlChainSuperParameter(value, force);
 }
 
-EffectChainPointer EffectChainSlot::getOrCreateEffectChain(
-        EffectsManager* pEffectsManager) {
-    if (!m_pEffectChain) {
-        EffectChainPointer pEffectChain(
-                new EffectChain(pEffectsManager, QString()));
-        //: Name for an empty effect chain, that is created after eject
-        pEffectChain->setName(tr("Empty Chain"));
-        loadEffectChainToSlot(pEffectChain);
-        pEffectChain->addToEngine(m_pEffectRack->getEngineEffectRack(), m_iChainSlotNumber);
-        pEffectChain->updateEngineState();
-        updateRoutingSwitches();
+void EffectChainSlot::setSuperParameterDefaultValue(double value) {
+    m_pControlChainSuperParameter->setDefaultValue(value);
+}
+
+void EffectChainSlot::slotChainEffectChanged(unsigned int effectSlotNumber) {
+    qDebug() << debugString() << "slotChainEffectChanged" << effectSlotNumber;
+    EffectSlotPointer pSlot;
+    EffectPointer pEffect;
+
+    if (m_effects.size() > m_slots.size()) {
+        qWarning() << debugString() << "has too few slots for effect";
     }
-    return m_pEffectChain;
+    if (effectSlotNumber < (unsigned) m_slots.size()) {
+        pSlot = m_slots.at(effectSlotNumber);
+    }
+    if (effectSlotNumber < (unsigned) m_effects.size()) {
+        pEffect = m_effects.at(effectSlotNumber);
+    }
+    if (pSlot != nullptr) {
+        pSlot->loadEffect(pEffect, m_pEffectRack->isAdoptMetaknobValueEnabled());
+    }
+
+    m_pControlNumEffects->forceSet(math_min(
+            static_cast<unsigned int>(m_slots.size()),
+            numEffects()));
```

Comment:
This is the only place the `numEffects` function is used. Remove that function and replace it with simply `m_effects.size()`.